Code/GraphMol/MolStandardize/Tautomer.cpp

```diff
@@ -12,57 +12,185 @@
 #include <GraphMol/SmilesParse/SmilesParse.h>
 #include <GraphMol/SmilesParse/SmilesWrite.h>
+#include <GraphMol/Substruct/SubstructMatch.h>
+#include <boost/dynamic_bitset.hpp>
 #include <algorithm>
+#include <limits>
+
+#include <boost/flyweight.hpp>
+#include <boost/flyweight/key_value.hpp>
+#include <boost/flyweight/no_tracking.hpp>
 
 using namespace RDKit;
 
 namespace RDKit {
 
 namespace MolStandardize {
 
+namespace detail {
+std::vector<std::pair<unsigned int, unsigned int>> pairwise(
+    const std::vector<int> vect) {
+  std::vector<std::pair<unsigned int, unsigned int>> pvect;
+  for (size_t i = 0; i < vect.size() - 1; ++i) {
+    std::pair<unsigned int, unsigned int> p =
+        std::pair<unsigned int, unsigned int>(vect[i], vect[i + 1]);
+    pvect.push_back(p);
+  }
+  return pvect;
+}
+
+}  // namespace detail
+
+namespace TautomerScoringFunctions {
+int scoreRings(const ROMol &mol) {
+  int score = 0;
+  auto ringInfo = mol.getRingInfo();
+  std::unique_ptr<ROMol> cp;
+  if (!ringInfo->isInitialized()) {
+    cp.reset(new ROMol(mol));
+    MolOps::symmetrizeSSSR(*cp);
+    ringInfo = cp->getRingInfo();
+  }
+  boost::dynamic_bitset<> isArom(mol.getNumBonds());
+  boost::dynamic_bitset<> bothCarbon(mol.getNumBonds());
+  for (const auto &bnd : mol.bonds()) {
+    if (bnd->getIsAromatic()) {
+      isArom.set(bnd->getIdx());
+      if (bnd->getBeginAtom()->getAtomicNum() == 6 &&
+          bnd->getEndAtom()->getAtomicNum() == 6) {
+        bothCarbon.set(bnd->getIdx());
+      }
+    }
+  }
+  for (const auto &bring : ringInfo->bondRings()) {
+    bool allC = true;
+    bool allAromatic = true;
+    for (const auto bidx : bring) {
+      if (!isArom[bidx]) {
+        allAromatic = false;
+        break;
+      }
+      if (!bothCarbon[bidx]) {
+        allC = false;
+      }
+    }
+    if (allAromatic) {
+      score += 100;
+      if (allC) {
+        score += 150;
+      }
+    }
+  }
+  return score;
+};
+
+struct smarts_mol_holder {
+  std::string d_smarts;
+  std::shared_ptr<ROMol> dp_mol;
+  smarts_mol_holder(const std::string &smarts) : d_smarts(smarts) {
+    dp_mol.reset(SmartsToMol(smarts));
+  }
+};
+
+typedef boost::flyweight<
+    boost::flyweights::key_value<std::string, smarts_mol_holder>,
+    boost::flyweights::no_tracking>
+    smarts_mol_flyweight;
+
+struct SubstructTerm {
+  std::string name;
+  std::string smarts;
+  int score;
+  std::shared_ptr<ROMol> matcher;
+  SubstructTerm(const std::string &aname, const std::string &asmarts,
+                int ascore)
+      : name(aname), smarts(asmarts), score(ascore) {
+    matcher = smarts_mol_flyweight(smarts).get().dp_mol;
+  };
+};
+
+int scoreSubstructs(const ROMol &mol) {
+  // a note on efficiency here: we'll construct the SubstructTerm objects here
+  // repeatedly, but the SMARTS parsing for each entry will only be done once
+  // since we're using the boost::flyweights above to cache them
+  const std::vector<SubstructTerm> substructureTerms{
```

Comment:
Part of me thinks that this structs + score should be passed in to be easier to modify.

Reply:
It looks like you've captured this in the score func though.

Reply:
It's also tested at match time (line 130)

Reply:
I thought about adding that option, but then figured it's more straightforward from the API perspective to just leave it out since the user can always provide their own scoring function.